apis/adc/src/lib.rs

```diff
@@ -0,0 +1,99 @@
+#![no_std]
+
+use core::cell::Cell;
+use libtock_platform::{
+    share, subscribe::OneId, DefaultConfig, ErrorCode, Subscribe, Syscalls, Upcall,
+};
+
+pub struct Adc<S: Syscalls>(S);
+
+impl<S: Syscalls> Adc<S> {
+    /// Returns Ok() if the driver was present.This does not necessarily mean
+    /// that the driver is working.
+    pub fn exists() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, EXISTS, 0, 0).to_result()
+    }
+
+    pub fn read_single_sample() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, SINGLE_SAMPLE, 0, 0).to_result()
+    }
+
+    // Register a listener to be called when the ADC conversion is finished
+    pub fn register_listener<'share, F: Fn(u16)>(
+        listener: &'share ADCListener<F>,
+        subscribe: share::Handle<Subscribe<'share, S, DRIVER_NUM, 0>>,
+    ) -> Result<(), ErrorCode> {
+        S::subscribe::<_, _, DefaultConfig, DRIVER_NUM, 0>(subscribe, listener)
+    }
+
+    /// Unregister the events listener
+    pub fn unregister_listener() {
+        S::unsubscribe(DRIVER_NUM, 0)
+    }
+
+    /// Initiates a synchronous ADC conversion
+    /// Returns the converted ADC value or an error
+    pub fn read_single_sample_sync() -> Result<u16, ErrorCode> {
+        let sample: Cell<Option<u16>> = Cell::new(None);
+        let listener = ADCListener(|adc_val| {
+            sample.set(Some(adc_val));
+        });
+        share::scope(|subscribe| {
+            if let Ok(()) = Self::register_listener(&listener, subscribe) {
+                if let Ok(()) = Self::read_single_sample() {
+                    while sample.get().is_none() {
+                        S::yield_wait();
+                    }
+                }
+            }
```

Comment:
Use the ?, similar to https://github.com/DanutAldea/libtock-rs/blob/b70eb0cbe1766a37585472235e77f755a7c12786/apis/ambient_light/src/lib.rs#L45-L46